analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs

```diff
@@ -0,0 +1,171 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ParameterTypeShouldMatchRouteTypeConstraint : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6800";
+    private const string MessageFormat = "{0}";
+    private const string MessageNoSecondaryLocationFormat = "Parameter type '{0}' does not match route parameter type constraint '{1}' in route '{2}'.";
+    private const string MessageWithSecondaryLocationFormat = "Parameter type '{0}' does not match route parameter type constraint.";
+    private const string SecondaryMessageFormat = "This route parameter has a '{0}' type constraint.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    private static readonly Dictionary<string, string> ConstraintMapping = new(StringComparer.InvariantCultureIgnoreCase)
+    {
+        { "bool", nameof(Boolean) },
+        { "datetime", nameof(DateTime) },
+        { "decimal", nameof(Decimal) },
+        { "double", nameof(Double) },
+        { "float", nameof(Single) },
+        { "guid", nameof(Guid) },
+        { "int", nameof(Int32) },
+        { "long", nameof(Int64) },
+        { "string", nameof(String) }
+    };
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(cc =>
+        {
+            // If we are not in a Blazor project, we don't need to register for lambda expressions.
+            if (cc.Compilation.GetTypeByMetadataName(KnownType.Microsoft_AspNetCore_Components_RouteAttribute) is null)
+            {
+                return;
+            }
+
+            cc.RegisterNodeAction(c =>
+            {
+                var node = (ClassDeclarationSyntax)c.Node;
+
+                foreach (var property in GetPropertyTypeMismatches(node, c.SemanticModel))
+                {
+                    var additionalLocations = property.RouteParamLocation is not null
+                        ? new[] { property.RouteParamLocation }
+                        : Array.Empty<Location>();
+                    var message = property.RouteParamLocation is not null
+                        ? string.Format(MessageWithSecondaryLocationFormat, property.Type.GetName())
+                        : string.Format(MessageNoSecondaryLocationFormat, property.Type.GetName(), property.ConstraintType, property.Route);
+
+                    c.ReportIssue(Diagnostic.Create(Rule,
+                        property.Type.GetLocation(),
```

Comment:
The property might be declared in a different file (e.g. base class). I'm not sure but I think Roslyn wasn't happy when the location of the report was in another syntax tree than the one where `c.Node` is part of. Do you have a test for such a case? If it doesn't work, you have to swap primary and secondary locations.

Reply:
I added one test to check this behavior when the route is defined in one file and the property in another.